### PR TITLE
Make .depend files generated for C sources more portable

### PR DESCRIPTION
**Pull request by shindere**
This implements a suggestion provided by @xavierleroy in
https://github.com/ocaml/ocaml/pull/941#issuecomment-266047690

### CODE REVIEW

**Comment by dra27** [comment]
Looks good to me - is this is a good time to address the dependencies in win32unix (which appear not to include C files at all?)

**Comment by shindere** [comment]
Many thanks for both your prompt response and your suggestion, @dra27!
I believe this dependency problem should indeed be fixed, yes.
Whether this should happen in the present PR or in a separate one I don't
know. It may go in the present one which is not that big...

**Comment by shindere** [comment]
Updated to also take into account the dependencies of the C files
of the win32unix library, thanks again for the reminder, @dra27.

**Comment by xavierleroy** [comment]
In favor!  (unsurprisingly)

**Comment by xavierleroy** [comment]
Travis fails because of no change entry, so I added the no-change-entry-needed label so that we can make some progress.  I'm more worried by the AppVeyor failure, which looks like a real problem.  Please check.

**Comment by shindere** [comment]
Xavier Leroy (2017/02/18 08:28 -0800):
> Travis fails because of no change entry, so I added the
> no-change-entry-needed label so that we can make some progress.  I'm
> more worried by the AppVeyor failure, which looks like a real problem.
> Please check.

So what I understand is that Travis fails because the PR odes not modify
the testsuite. I don't know what to do about that.

AppVeyor seems to fail, too, because "Files unixLables.cmo and unix.cmo
make inconsistent assumptions over interface Unix".
I don't know how to deal with this, either. Isn't it justa a matter of
really cleaning the build environment and running the build again?
Perhaps the appveyor scriptshould make sure to do a make distclean
before starting to compile?

**Comment by xavierleroy** [comment]
> "Files unixLables.cmo and unix.cmo make inconsistent assumptions over interface Unix". I don't know how to deal with this, either. 

This often comes from missing dependencies causing some files not to be remade.  It so happens that you're in the process of changing dependencies...



> Isn't it justa a matter of really cleaning the build environment and running the build again? Perhaps the appveyor scriptshould make sure to do a make distclean before starting to compile?

No, because in this case we would find fewer 'missing dependencies' issues.

**Comment by dra27** [comment]
@shindere - the testsuite modifying test is informational only (it's permitted to fail); the problem is the Changes file. We're frequently hitting GitHub API limits at the moment. I looked into it a few weeks ago - I've been meaning to talk to @avsm about setting up a mirror of the GPR labels on non-rate-limited infrastructure.

**Comment by shindere** [comment]
Xavier Leroy (2017/02/20 00:48 -0800):
> > "Files unixLables.cmo and unix.cmo make inconsistent assumptions over interface Unix". I don't know how to deal with this, either. 
> 
> This often comes from missing dependencies causing some files not to
> be remade.  It so happens that you're in the process of changing
> dependencies...

My bad, there were indeed missing dependencies...

+unix.cmo : unix.cmi
+unix.cmx : unix.cmi

PR updated and rebased, things should work much better now, sorry for
the noise.

**Comment by shindere** [comment]
Just rebased this PR on trunk.

Is there anything missing for it to be merged?

**Comment by dra27** [comment]
I have one further thought (and sorry for its being late in the process): you have made the Makefiles consistent in **not** having a depend target for msvc (cf. otherlibs/str/Makefile which didn't and byterun/Makefile which did beforehand) but I wonder if this is the wrong path. At present, make alldepend fails on msvc with a cryptic error message because the depend target isn't defined.

In each case where there is

```make
ifneq "$(TOOLCHAIN)" "msvc"
```

would it better to have:

```make
ifeq "$(TOOLCHAIN)" "msvc"
$(error Dependencies cannot be regenerated using the MSVC ports)
else
```

Have you run a make alldepend since rebasing? I went through the differences caused by running make alldepend on mingw64. I think a lot may be that make alldepend was just missing, but here's the complete list:

asmrun/afl{,.p,.d,.i}.o
	Order of dependencies changes (I'm not sure why this isn't stable - perhaps one make alldepend is enough?)

asmrun/str{,.p,.d,.i}.o
	Some dependencies are listed twice, and several more on memory.h, major_gc.h, minor_gc.h, address_class.h, freelist.h appeared

asmrun/unix{,.p,.d,.i}.o **DISAPPEARS**

asmrun/win32{,.p,.d,.i}.o **ADDED**

Similar effect with str.o and afl.o in byterun/

byterun/win32.o dependencies gain flexdll.h (non-critical)

debugger/loadprinter.cmo gains dependency on ../parsing/location.cmi
debugger/loadprinter.cmx gains dependency on ../parsing/location.cmx

otherlibs/bigarray/.depend
	Unix dependencies re-map from unix->win32unix

I wonder if a few of these are worthy of dealing with:

 - asmrun's depend target should always consider win32.o and unix.o
 - Perhaps a sed trick can map the dependency on unixsupport.h to the correct directory (UNIX_OR_WIN32 or something in config/Makefile, I think?)
 - The dependency on flexdll.h is awkward, but it should certainly be filtered in some way, as it would be problematic

**Comment by shindere** [comment]
Thanks a lot for your feedback, David.

David Allsopp (2017/03/01 03:56 -0800):
> I have one further thought (and sorry for its being late in the process):

no problem! Later is better than never!

> you have made the Makefiles consistent in **not** having a depend
> target for msvc (cf. otherlibs/str/Makefile which didn't and
> byterun/Makefile which did beforehand)

Indeed

> but I wonder if this is the wrong path. At present, make alldepend fails on msvc with a cryptic error message because the depend target isn't defined.
> 
> In each case where there is
> 
> ```make
> ifneq "$(TOOLCHAIN)" "msvc"
> ```
> 
> would it better to have:
> 
> ```make
> ifeq "$(TOOLCHAIN)" "msvc"
> $(error Dependencies cannot be regenerated using the MSVC ports)
> else
> ```

How about having this code in the root Makefile for the alldepend rule
and simply refusing to run it completely on MSVC ports?

> Have you run a make alldepend since rebasing?

Nope. Should I have done so? If yes, I have to say I do not understand
why, sorry.

> I went through the differences caused by running make alldepend on mingw64. I think a lot may be that make alldepend was just missing, but here's the complete list:

[...]

Thanks! Should these things be fixed in this PR or would it make more
sense to fix them in another one?

**Comment by dra27** [comment]
To me, philosophically, it would be better for each Makefile which has a depend target to have a meaningful error about why it's not available, rather than the error - but I agree that just replacing alldepend in the root Makefile is indisputably simpler!

For running make alldepend, I'm not totally sure either: I think that these are to do with changes where the previous commits should themselves have run make alldepend.

Similarly with the mingw64 differences, I'm not sure if this is strictly for here, or just worth fixing at the same time. I guess it's not creating blocking problems, but the lack of win32.o dependencies, etc. isn't great?

**Comment by shindere** [comment]
David Allsopp (2017/03/01 05:54 -0800):
> To me, philosophically, it would be better for each Makefile which has
> a depend target to have a meaningful error about why it's not
> available, rather than the error - but I agree that just replacing
> alldepend in the root Makefile is indisputably simpler!

Well, to be honest, I think it is not really worth the effort, because
the idea is to change the build system so that the makefiles in
subdirectories get included in the root makefile rather than invoked. So
ultimately all the dependencies will be computed by the root makefile
and the sub-makefiles would just have default targets that call the root
makefile, to keep the possibility of running make in the subdirectories
as is done now.

> For running make alldepend, I'm not totally sure either: I think that
> these are to do with changes where the previous commits should
> themselves have run make alldepend.

That's my understanding, too.

> Similarly with the mingw64 differences, I'm not sure if this is
> strictly for here, or just worth fixing at the same time. I guess it's
> not creating blocking problems, but the lack of win32.o dependencies,
> etc. isn't great?

Yeah, in theory I agree. Practically speaking, though, what this means
is that the dependencies should always be generated on a system where
win32.c is taken into account. In particular, this in my understanding
would mean that we should all refrain frm generating dependencies under
Linux, unless of course the Unix build system is tweaked so that make
depend also takes the Windows-specific
c files into account. Shall we go for that?

**Comment by shindere** [comment]
@dra27: the latest version of the PR tries to address your comments in
https://github.com/ocaml/ocaml/pull/1047#issuecomment-283321574

In particular, `make [all]depend` now fails with a clear error message on
msvc.

The PR also makes an effort to compute dependencies for files although
they would not be compiled on the platform where `make alldepend` is run,
e.g. on a given platform not all libraries under otherlibs are compiled
but with this PR `make alldepend` will compute dependencies even for the
libraries that don't get compiled on this platform.

The last commit of the PR updates all the dependencies with the result
of make alldepend run on a Linux amd64 platform.

One thing to keep in mind is that `make alldepend` will continue to
produce platform-specific results, for the following reasons:

- The architecture-specific files have different dependencies but the
  build system does not distinguish them from each other
  (could be fixed)

- The order in which dependencies are computed varies slightly from
  platfom to platfom and so does their presentaiton. This of course has
  no effect on the dependency graph itself but introduces some noise in
  the diffs.
  (could likely be fixed)

- More importantly: completely accurate dependencies can be known only
  at compile time. For instance, foo.c may depend on a.h or b.h
  depending on some preprocessor macro whose value is known for sure
  only at compile time.

To sum up: as long as the depencies will be stored statically in the
repository as is currently done, it will not be possible to have
completely accurate dependencies on all platforms. Given the limitations
inherent to this static approach, I'd suggest to not put too much energy in
improving it and just fix it when it stops working.

My personal preference would be that the dependencies are handled in a
completely dynamic way, but given that this problem is not really a
blocker (things more or less work, although not completely
satisfactory), I believe this is not really a high priority, at the moment.
